.ci/jenkins-execute-script.py

```diff
@@ -0,0 +1,89 @@
+repo_url = os.environ['REPO_URL']  if os.environ.get('REPO_URL') != None else 'https://github.com/projectatomic/minishift-centos-iso.git'
+branch = os.environ['BRANCH']  if os.environ.get('BRANCH') != None else 'master'
+
+def execute_on_host( host, cmd, error_message ):
+    # build command to execute install and test commands via ssh
+    ssh_cmd  = "ssh -t -t "
+    ssh_cmd += "-o UserKnownHostsFile=/dev/null "
+    ssh_cmd += "-o StrictHostKeyChecking=no "
+    ssh_cmd += "root@%s " % (host)
+
+    cmd = '%s "%s"' % (ssh_cmd, cmd)
+    print "Executing: %s" % (cmd)
+    exit_code = subprocess.call(cmd, shell=True)
+    if exit_code != 0 : sys.exit(error_message)
+    return
+
+def prepare_pull_request_build(host):
+    pr_branch = os.environ['ghprbSourceBranch']
+    pr_author_repo = os.environ['ghprbAuthorRepoGitUrl']
+
+    branch_cmd  = 'cd minishift-centos-iso && '
+    branch_cmd += "git checkout -b %s" % (pr_branch)
+    execute_on_host(host, branch_cmd, "Unable to create branch for pull request build")
+
+    pull_cmd  = 'cd minishift-centos-iso && '
+    pull_cmd += "git pull --no-edit %s %s " % (pr_author_repo, pr_branch)
+    execute_on_host(host, pull_cmd, "Unable to pull pull request")
+    return
+
+for host in data['hosts']:
+    # run the Ansible playbook
+    ansible_cmd  = 'yum -y install git epel-release livecd-tools && '
+    ansible_cmd += 'yum -y install ansible1.9 && '
+    ansible_cmd += 'git clone %s && ' % repo_url
+    ansible_cmd += 'cd minishift-centos-iso && '
+    ansible_cmd += 'git checkout %s && ' % branch
+    ansible_cmd += 'cd .ci/ansible && '
+    ansible_cmd += 'ANSIBLE_NOCOLOR=1 ansible-playbook site.yml'
+    execute_on_host(host, ansible_cmd, "Ansible playbook failed")
+
+    # if we deal with a pull request build we need to prepare the source
+    if os.environ.get('ghprbPullId') != None:
+        prepare_pull_request_build(host)
+
+    # setup kvm driver
+    setup_cmd = 'curl -L https://github.com/dhiltgen/docker-machine-kvm/releases/download/v0.7.0/docker-machine-driver-kvm > /usr/local/bin/docker-machine-driver-kvm && '
+    setup_cmd += 'chmod +x /usr/local/bin/docker-machine-driver-kvm'
+    execute_on_host(host, setup_cmd, "Failed to setup kvm driver")
+
+    # copy the docker-machine binary
+    setup_cmd  = 'cd minishift-centos-iso && mkdir -p bin && cd bin && '
+    setup_cmd += 'curl -L https://github.com/docker/machine/releases/download/v0.8.2/docker-machine-`uname -s`-`uname -m` > docker-machine && '
+    setup_cmd += 'chmod +x docker-machine'
+    execute_on_host(host, setup_cmd, "Failed to setup docker-machine")
+
+    # run build and features
+    build_cmd  = 'cd minishift-centos-iso && '
+    build_cmd += 'make centos_iso && '
+    build_cmd += 'make test'
+    execute_on_host(host, build_cmd, "Tests failures")
+
+    done_nodes_url = "%s/Node/done?key=%s&sside=%s" % (url_base, api_key, data['ssid'])
+    print urllib.urlopen(done_nodes_url)
```

Comment:
there are a lot of conditions where this line wont even get executed - how are you trapping for orphaned machine nodes ?